int64 up_px
int64 down_px
float32 slope
bool hasSlope